server/pulp/server/controllers/content.py

```diff
@@ -128,113 +137,392 @@ def refresh_content_source(content_source_id=None):
+    :rtype:  generator of pulp.server.db.model.FileContentUnit
+    """
+    for deferred_download in DeferredDownload.objects.filter():
+        try:
+            unit_model = plugin_api.get_unit_model_by_id(deferred_download.unit_type_id)
+            unit = unit_model.objects.filter(id=deferred_download.unit_id).get()
+            yield unit
+        except ModelNotFound:
+            _logger.error(_('Unable to find the model object for the {type} type.').format(
+                type=deferred_download.unit_type_id))
+        except DoesNotExist:
+            # This is normal if the content unit in question has been purged during an
+            # orphan cleanup.
+            _logger.debug(_('Unable to find the {type}:{id} content unit.').format(
+                type=deferred_download.unit_type_id, id=deferred_download.unit_id))
+
+
+def _create_download_requests(content_units):
+    """
+    Make a list of Nectar DownloadRequests for the given content units using
+    the lazy catalog.
 
+    :param content_units: The content units to build a list of DownloadRequests for.
+    :type  content_units: list of pulp.server.db.model.FileContentUnit
 
-def _get_streamer_url(catalog_entry):
+    :return: A list of DownloadRequests; each request includes a ``data``
+             instance variable which is a dict containing the FileContentUnit,
+             the list of files in the unit, and the downloaded file's storage
+             path.
+    :rtype:  list of nectar.request.DownloadRequest
     """
-    Translate a content unit into a URL where the content unit is cached.
+    requests = []
+    working_dir = get_working_directory()
+    signing_key = Key.load(pulp_conf.get('authentication', 'rsa_key'))
+
+    for content_unit in content_units:
+        # All files in the unit; every request for a unit has a reference to this dict.
+        unit_files = {}
+        unit_working_dir = os.path.join(working_dir, content_unit.id)
+        for file_path in content_unit.list_files():
+            qs = LazyCatalogEntry.objects.filter(
+                unit_id=content_unit.id,
+                unit_type_id=content_unit.type_id,
+                path=file_path
+            )
+            catalog_entry = qs.order_by('revision').first()
+            signed_url = _get_streamer_url(catalog_entry, signing_key)
+
+            temporary_destination = os.path.join(
+                unit_working_dir,
+                os.path.basename(catalog_entry.path)
+            )
+            mkdir(unit_working_dir)
+            unit_files[temporary_destination] = {
+                CATALOG_ENTRY: catalog_entry,
+                PATH_DOWNLOADED: None,
+            }
+            request = DownloadRequest(signed_url, temporary_destination)
+            request.data = {
+                CONTENT_UNIT: content_unit,
+                UNIT_FILES: unit_files,
+            }
+            requests.append(request)
```

Comment:
This is a great opportunity for making a generator and yielding requests one at a time. Each iteration of this loop is relatively expensive, and you can imagine that it might take substantial time if there are tens of thousands of content units to work through. That could also take a lot of RAM, especially for RPMs.